### DISCUSSION TITLE
serveur sending websocket message before shutdown

### DISCUSSION THREAD

**Discussion by sophiedegran** [miguelgrinberg/microdot#116] [Q&A]
Miguel

All work fine on raspberry pico it's amazing project!

I manage a websocket list like this:
[token_key, ip, sock, web_socket_key]

@app.route ('/update')
@with_websocket

def update (request, ws):
token_key = random value 
ip = request.client_addr
sock = request.sock
web_socket_key = request.headers.get ('Sec-WebSocket-Key', '')

How can I send to the list the message ? I suppose i have to create a request in microdot with specifiq websocket ?
Please help me! It's really hard for me to do that !!!

Regards

**Comment by miguelgrinberg** [comment]
I'm sorry but I don't understand what you want to do. You mentioned a list, but there is no list in your code. And what is this shutdown reference in the title? I think you are not giving me the full context.

**Comment by sophiedegran** [comment]
Ok so sorry

So I manage a list as explain on each websocket open by  clients
_websocket_list.append [token_key, ip, sock, web_socket_key]

@app.route ('/shutdown')
def shutdown (request):
for x in range(len(_websocket_list) ):
  message = 'Serveur will shutdown in 5 sec'
  -->here would like to send ...

**Comment by miguelgrinberg** [comment]
Something like this maybe?

```python
for ws in _websocket_list:
    ws[2].send('Serveur will shutdown in 5 sec')
```

Here I'm using the 3rd element in each of your clients, which is the `ws` object that was passed to each client handler.

**Comment by sophiedegran** [comment]
Miguel,

Ugrade to 1.2.1 many thanks
I take long time to answer you because i want to investigate what was the problem with doing
ws [2].send ('Message') 
If i do that I've immediatly :
microdot.py line 1113 in dispatch_request
res = f (req, **req.url_args)
TypeError: a bytes-like object is required not str
500 --> Internal server error

So i try to force  byte ws [2].send (b'Message')
and all stations in the list received something but done socket close ...
On all clients console Edge: 
(index):558 Websocket connection to
'ws://127.0.0.1:5000/update' failed: One or more reserved bits are on reserved1=1, reserved2=1, reserved3=1 
and after got
socket close

Voila, voila... 
1- can't explain why string (i try json data also) not accepted
2- remenber the page is doing on /update websocket and i send from /shutdown perharps a security reason for that point

Regards

**Comment by miguelgrinberg** [comment]
> 1. can't explain why string (i try json data also) not accepted

I think you have not included the complete error from MicroPython, with the full stack trace. That makes it impossible for me to understand what you get the error.

> 2- remenber the page is doing on /update websocket and i send from /shutdown perharps a security reason for that point

Do you have any antivirus? Looks like this error could be related to aggressive protocol filtering. Try disabling to see if the error goes away. Reference: https://forum.eset.com/topic/8355-protocol-filtering-breaks-websockets-after-update-at-5052016/

**Comment by sophiedegran** [comment]
1-
Traceback (most recent call last):
File "C:\sophie\venv\lib\site-packages\microdot.py", line 1113, in
dispatch_request
   res = f (req, **req.url_args)
File "C:\sophie\microdot\microdot_socket.py", line 638, in shutdown
  ws[2].send ('test send')
TypeError: a bytes-like object is required, not 'str'
Get /shutdown 500
Get /update 200

(for 2 clients ...)


2- will look at that thanks and tell you if improvement

Regards



Le sam. 10 déc. 2022 12:05, Miguel Grinberg ***@***.***> a
écrit :

**Comment by miguelgrinberg** [comment]
What is the point of storing a plain socket? From your code, you are not storing a WebSocket instance, you are storing the raw socket:

```
def update (request, ws):
token_key = random value
ip = request.client_addr
sock = request.sock     # <--- this is a plain socket, not WebSocket
web_socket_key = request.headers.get ('Sec-WebSocket-Key', '')
```

I also don't understand why you store all the other things, but this socket should be the WebSocket instance, I think:

```
sock = ws
```

**Comment by sophiedegran** [comment]
Yes all work perfectly !
I was doing that because don't understand how to use ws,  now it's very clear. I have only to have a simple list with [station_A, ws_station] .
It 's a real improvement for me as we can do micro chat, interactive draw, etc.. ☺
Many many Thanks
